src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -349,261 +179,78 @@ public virtual void GetPolicies()
         protected JsonSerializerOptions Options { get; set; } = null!; // initialized in Init method
 
-        protected abstract void OnRead(ref ReadStack state, ref Utf8JsonReader reader);
-        protected abstract void OnReadEnumerable(ref ReadStack state, ref Utf8JsonReader reader);
-        protected abstract void OnWrite(ref WriteStackFrame current, Utf8JsonWriter writer);
-        protected virtual void OnWriteDictionary(ref WriteStackFrame current, Utf8JsonWriter writer) { }
-        protected abstract void OnWriteEnumerable(ref WriteStackFrame current, Utf8JsonWriter writer);
-
-        public Type ParentClassType { get; private set; } = null!;
-
-        public PropertyInfo? PropertyInfo { get; private set; }
-
-        public void Read(JsonTokenType tokenType, ref ReadStack state, ref Utf8JsonReader reader)
+        public bool ReadJsonAndAddExtensionProperty(object obj, ref ReadStack state, ref Utf8JsonReader reader)
         {
-            Debug.Assert(ShouldDeserialize);
+            object propValue = GetValueAsObject(obj)!;
 
-            JsonPropertyInfo? propertyInfo;
-            JsonClassInfo? elementClassInfo = ElementClassInfo;
-            if (elementClassInfo != null && (propertyInfo = elementClassInfo.PolicyProperty) != null)
+            IDictionary<string, object?>? dictionaryObject = propValue as IDictionary<string, object?>;
+            if (dictionaryObject != null)
             {
-                if (!state.Current.CollectionPropertyInitialized)
+                // Handle case where extension property is object-based.
+
+                if (reader.TokenType == JsonTokenType.Null)
                 {
-                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(propertyInfo.RuntimePropertyType);
+                    // A null JSON value is treated as a null object reference.
+                    dictionaryObject[state.Current.JsonPropertyNameAsString!] = null;
                 }
+                else
+                {
+                    JsonConverter<object> converter = (JsonConverter<object>)
+                        state.Current.JsonPropertyInfo!.RuntimeClassInfo.ElementClassInfo!.PolicyProperty!.ConverterBase;
 
-                // Forward the setter to the value-based JsonPropertyInfo.
-                propertyInfo.ReadEnumerable(tokenType, ref state, ref reader);
-            }
-            // For performance on release build, don't verify converter correctness for internal converters.
-            else if (HasInternalConverter)
-            {
-#if DEBUG
-                JsonTokenType originalTokenType = reader.TokenType;
-                int originalDepth = reader.CurrentDepth;
-                long originalBytesConsumed = reader.BytesConsumed;
-#endif
-
-                OnRead(ref state, ref reader);
+                    if (!converter.TryRead(ref reader, typeof(JsonElement), Options, ref state, out object value))
```

Comment:
This should return a nullable object. Also the converter should be `JsonConverter<object?>`
```suggestion
                    if (!converter.TryRead(ref reader, typeof(JsonElement), Options, ref state, out object? value))
```